hack/generator/pkg/codegen/pipeline_add_cross_resource_references.go

```diff
@@ -0,0 +1,181 @@
+}
+
+type crossResourceReferenceChecker func(typeName astmodel.TypeName, prop *astmodel.PropertyDefinition) bool
+
+type crossResourceReferenceTypeVisitor struct {
+	astmodel.TypeVisitor
+	// referenceChecker is a function describing what a cross resource reference looks like. It is overridable so that
+	// we can use a more simplistic criteria for tests.
+	isPropertyAnARMReference crossResourceReferenceChecker
+}
+
+func makeCrossResourceReferenceTypeVisitor(idFactory astmodel.IdentifierFactory, referenceChecker crossResourceReferenceChecker) crossResourceReferenceTypeVisitor {
+	visitor := crossResourceReferenceTypeVisitor{
+		isPropertyAnARMReference: referenceChecker,
+	}
+
+	transformResourceReferenceProperties := func(this *astmodel.TypeVisitor, it *astmodel.ObjectType, ctx interface{}) (astmodel.Type, error) {
+		typeName := ctx.(astmodel.TypeName)
+
+		var newProps []*astmodel.PropertyDefinition
+		for _, prop := range it.Properties() {
+			if visitor.isPropertyAnARMReference(typeName, prop) {
+				klog.V(4).Infof("Transforming \"%s.%s\" field into genruntime.ResourceReference", typeName, prop.PropertyName())
+				originalName := string(prop.PropertyName())
+				prop = makeResourceReferenceProperty(idFactory, prop)
+
+				// TODO: We could pass this information forward some other way?
+				// Add tag so that we remember what this field was before
+				prop = prop.WithTag(astmodel.ARMReferenceTag, originalName)
+			}
+
+			newProps = append(newProps, prop)
+		}
+
+		it = it.WithoutProperties()
+		result := it.WithProperties(newProps...)
+		return result, nil
+	}
+
+	visitor.TypeVisitor = astmodel.TypeVisitorBuilder{
+		VisitObjectType: transformResourceReferenceProperties,
+	}.Build()
+
+	return visitor
+}
+
+// doesPropertyLookLikeARMReference uses a simple heuristic to determine if a property looks like it might be an ARM reference.
+// This can be used for logging/reporting purposes to discover references which we missed.
+func doesPropertyLookLikeARMReference(prop *astmodel.PropertyDefinition) bool {
+	// The property must be a string or optional string
+	isString := prop.PropertyType().Equals(astmodel.StringType)
+	isOptionalString := prop.PropertyType().Equals(astmodel.NewOptionalType(astmodel.StringType))
+	if !isString && !isOptionalString {
+		return false
+	}
+
+	hasMatchingDescription := armIDDescriptionRegex.MatchString(prop.Description())
+	namedID := prop.HasName("Id")
+	if hasMatchingDescription || namedID {
+		return true
+	}
+
+	return false
+}
+
+func makeResourceReferenceProperty(idFactory astmodel.IdentifierFactory, existing *astmodel.PropertyDefinition) *astmodel.PropertyDefinition {
+
+	var referencePropertyName string
+	// Special case for "Id" which is quite common in the specs and also contains no usage information we're afraid to lose (as opposed to
+	// something like "vnet" what says what the reference is pointing to and we don't want to overwrite to something generic).
+	if existing.PropertyName() == "Id" {
```

Comment:
Can you expand on why you need to rename the property? I guess it's because a field called `Id` but containing a reference is awkward?

Reply:
If the property wasn't renamed, could you do without the tag by triggering the reference handling in the ARM conversion pipeline off the `ResourceReference` type (like the FindResourceReferences reflect visitor)?

Reply:
Can expand.

It's just because `id: ComplexStructure` is awkward and really common. Ends in ID is _also_ really common though so maybe should handle that here too. (change it to ends in reference)

Reply:
Also deal with `Id` suffix now, added a golden file test for `Id` suffix as well.